terraform-modules/teleport-aws-discovery/main.tf

```diff
@@ -0,0 +1,287 @@
+locals {
+  create                = var.create
+  discover_organization = false # TODO(gavin): impl org discovery
+  name_prefix = (
+    var.name_prefix != ""
+    ? "${trimsuffix(var.name_prefix, "-")}-"
+    : ""
+  )
+  tags = merge(var.tags, {
+    "teleport.dev/cluster"     = local.teleport_cluster_name
+    "teleport.dev/integration" = local.teleport_integration_name
+    # this is the origin we set for resources created by the AWS OIDC integration web UI wizard.
+    "teleport.dev/origin" = "integration_awsoidc"
+  })
+  teleport_cluster_name      = var.teleport_cluster_name
+  teleport_proxy_public_addr = var.teleport_proxy_public_addr
+  teleport_proxy_public_url  = "https://${local.teleport_proxy_public_addr}"
+}
+
+################################################################################
+# AWS IAM OIDC Provider
+################################################################################
+
+locals {
+  create_aws_oidc_provider  = local.create
+  aws_iam_oidc_provider_aud = "discover.teleport"
+  # strip the port since AWS OIDC provider doesn't support port in the url
+  aws_iam_oidc_provider_url = replace(local.teleport_proxy_public_url, "/:[0-9]+.*/", "")
+  default_aws_resource_name = "teleport-discovery"
+}
+
+data "tls_certificate" "teleport_proxy" {
+  count = local.create_aws_oidc_provider ? 1 : 0
+
+  url = local.teleport_proxy_public_url
+}
+
+# Create an AWS OIDC Provider, so that the Teleport Discovery Service can use
+# OIDC to assume the discovery AWS IAM role.
+resource "aws_iam_openid_connect_provider" "teleport" {
+  count = local.create_aws_oidc_provider ? 1 : 0
+
+  url             = local.aws_iam_oidc_provider_url
+  client_id_list  = [local.aws_iam_oidc_provider_aud]
+  thumbprint_list = [data.tls_certificate.teleport_proxy[0].certificates[0].sha1_fingerprint]
+  tags            = local.tags
+}
+
+################################################################################
+# AWS IAM role for Teleport Discovery Service
+################################################################################
+
+locals {
+  aws_iam_oidc_provider_arn = try(
+    aws_iam_openid_connect_provider.teleport[0].arn,
+    "",
+  )
+  create_teleport_discovery_service_iam_role = local.create
+  teleport_discovery_service_iam_role_name = "${local.name_prefix}${coalesce(
+    var.teleport_discovery_service_iam_role_name,
+    local.default_aws_resource_name,
+  )}"
+  teleport_discovery_service_iam_role_trust_policy = <<EOF
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "Federated": "${local.aws_iam_oidc_provider_arn}"
+            },
+            "Action": "sts:AssumeRoleWithWebIdentity",
+            "Condition": {
+                "StringEquals": {
+                    "${local.teleport_cluster_name}:aud": "${local.aws_iam_oidc_provider_aud}"
+                }
```

Comment:
idk how it works with AWS stuff, but does this need to have `https://` anywhere in the issuer url?